### PR TITLE
use twitch_types for types from twitch

### PR DESCRIPTION
**Pull request by Emilgardis**

### CODE REVIEW

**Comment by Emilgardis** [comment]
For your consideration.

I think this is nice but interfacing with `twitch_types` is annoying sometimes since the experience is not a straight `str -> T` (since the idea is that you shouldn't be able to do mistakes like confusing usernames with display names etc). The braids do try to provide an as good experience as possible.

I've minimized deps in `twitch_types` (0), the `twitch_types/serde` feature uses `serde_derive`, but this crate already does aswell

File: deny.toml [hunk 1]
```diff
@@ -13,7 +13,8 @@ unlicensed = "deny"
 allow = [
     "MIT",
     "Apache-2.0",
-    "0BSD"
+    "0BSD",
+    "Unicode-DFS-2016"
```
Comment (Emilgardis) [inline deny.toml]: not sure why this didn't get flagged before

Comment (Emilgardis) [reply deny.toml]: ah, need to invoke `cargo deny --all-features check`, fixed in 3bfb0b51ebf9f3f663dbf5d132187d67ca4bacbf

**Comment by museun** [comment]
This is actually something I was going to ask you about.

What does `IntoCow` do? I assume `Cow::Borrowed("broadcaster".into())` is making a typed string from the `&str` via the `.into()` and then into a `Cow<'a, Typed>`

**Comment by Emilgardis** [comment]
`IntoCow` is defined [here](https://github.com/twitch-rs/twitch_types/blob/88db36d9dc98ad01451e1d9c9c9fbf5a3f731de6/src/lib.rs#L11) ([docs](https://docs.rs/twitch_types/0.4.0/twitch_types/trait.IntoCow.html)).

Does exactly that, via [this impl](https://docs.rs/twitch_types/0.4.0/twitch_types/trait.IntoCow.html#impl-IntoCow%3C%27a%2C%20R%3E-for-%26%27a%20str).

edit: the `From<Owned> for String` is defined [here](https://github.com/twitch-rs/twitch_types/blob/88db36d9dc98ad01451e1d9c9c9fbf5a3f731de6/src/macros.rs#L57-L62)

**Comment by museun** [comment]
Also related, do you have ownership of twitch-rs? We can move this crate under that organization if you'd approve that.

I'm planning on adding an enum interface to the types soon. So instead of struct + tag it'd have the option of just using a big enum. I'll think about your crate while I plan out the design of that and then decide on this change or not.

I'll keep the `as_typed_message` thing but also provide `as_enum` and `into_enum`.

**Comment by Emilgardis** [comment]
I sent an invitation to the org :)

**Comment by museun** [comment]
I plan on making a few more crates, `twitch_message_ws` (perhaps features for `tokio-tungstenite` and regular `tungstenite`) `twitch_message_tokio` and `twitch_message_tcp` (*names pending*). Which'll provide the actual i/o side to this. We can place them all under the twitch-rs umbrella.

Perhaps I'll even make an extensible framework like serenity or or twilight for writing bots using these (the tokio one, most likely.). 

That way your org also has a twitch chat set of crates. I'll have to check the available crates for oauth flow implementations to figure out which to use.

How should we move forward? I think wider integration is a great idea. If this is moved, then we can go ahead and approve this. As long as we have a way to get the untyped &str out then I don't see any problem.

**Comment by Emilgardis** [comment]
I think that sounds great! 

For oauth, I used to use `oauth2` but that one needed manual tinkering and didn't have a good client abstraction so I use [`twitch_oauth2`](https://github.com/twitch-rs/twitch_oauth2) which handles it all. [example](https://docs.rs/twitch_oauth2/0.11.0/twitch_oauth2/tokens/struct.UserTokenBuilder.html#examples)

**Comment by Emilgardis** [comment]
i'll rebase the branch

**Comment by museun** [comment]
Its the tests I added before the braid types were added. The tests don't care about what the data are, just the "shape". Its to test the lifetime conversion and that each type is covered. So whatever placeholder can be used instead of those `Cow::default()` calls

**Comment by museun** [comment]
@Emilgardis is this ready to be merged?

File: src/messages.rs [hunk 2]
```diff
@@ -55,3 +55,8 @@ pub use user_notice::{UserNotice, UserNoticeId};
 
 mod twitch_message;
 pub use self::twitch_message::TwitchMessage;
+
+pub use twitch_types::{
+    self as types, BadgeSetIdRef, ChatBadgeIdRef, DisplayNameRef, EmoteIdRef, EmoteSetIdRef,
+    IntoCow, MsgIdRef, UserIdRef, UserNameRef, UserType,
+};
```
Comment (museun) [inline src/messages.rs:62]: should this be re-exported like this? or should it just refer to the extern crate and re-export the crate at the root?

*edit* we can decide before the publish

**Review by museun — COMMENTED** [review]
This seems appropriate

**Pull request merged**